pkg/container/immset.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package container
+
+import (
+	"slices"
+
+	"golang.org/x/exp/constraints"
+)
+
+// ImmSet is an immutable set optimized for a smallish (1-1000) set of items.
+// Implemented as a sorted slice.
+type ImmSet[T any] struct {
+	xs  []T
+	cmp func(T, T) int
+	eq  func(T, T) bool
+}
+
+func NewImmSet[T constraints.Ordered](items ...T) ImmSet[T] {
+	return NewImmSetFunc[T](ordCmp[T], items...)
+}
+
+func NewImmSetFunc[T any](compare func(T, T) int, items ...T) ImmSet[T] {
+	s := ImmSet[T]{items, compare, cmpToEqual(compare)}
+	slices.SortFunc(s.xs, s.cmp)
+	s.xs = slices.CompactFunc(s.xs, s.eq)
+	return s
+}
+
+func (s ImmSet[T]) AsSlice() []T {
+	return s.xs
```

Comment:
Returning the underlying slice here will allow callers to modify it and potentially invalidate the invariant of the slice being sorted. Should this return a copy instead?

Reply:
That's true, but the whole point of adding the ImmSet is to have a fast implementation of an Immutable set. I would like to avoid additional copies and rely on the proper usage of the type.
What about adding a comment on this method to clarify that the caller is not allowed to modify the returned slice?

Reply:
Adding a comment to the method sounds good to me :+1: